### PR TITLE
Make TSUrlSchemeGet() return scheme implied by URL type when there is no explicit scheme.

### PR DESCRIPTION
**Pull request by ywkaras**
The original TSUrlSchemeGet() is renamed to TSUrlRawSchemeGet().

### CODE REVIEW

File: tests/tools/plugins/test_tsapi.cc [hunk 1]
```diff
@@ -28,14 +28,6 @@ Regression testing code for TS API.  Not comprehensive, hopefully will be built
...
 #include <tscpp/util/PostScript.h>
 
-// TSReleaseAssert() doesn't seem to produce any logging output for a debug build, so do both kinds of assert.
-//
-#define ALWAYS_ASSERT(EXPR) \
-  {                         \
-    TSAssert(EXPR);         \
-    TSReleaseAssert(EXPR);  \
-  }
-
```
Comment (ywkaras) [inline tests/tools/plugins/test_tsapi.cc:38]: Some unrelated cleanup, included in this PR.

File: tests/gold_tests/pluginTest/tsapi/log.gold [hunk 2]
```diff
@@ -1,14 +1,87 @@
...
 TSHttpTxnEffectiveUrlStringGet():  http://mYhOsT.teSt:SERVER_PORT/
+Client Request:
 TSHttpHdrEffectiveUrlBufGet():  http://myhost.test:SERVER_PORT/
+TSUrlSchemeGet():  http
+TSUrlRawSchemeGet():  http
+TSUrlPortGet():  80
+Global: event=TS_EVENT_HTTP_SEND_REQUEST_HDR
+TSHttpTxnEffectiveUrlStringGet():  http://127.0.0.1:SERVER_PORT/
+Request To Server:
+TSHttpHdrEffectiveUrlBufGet():  127.0.0.1:SERVER_PORT/
```
Comment (ywkaras) [inline tests/gold_tests/pluginTest/tsapi/log.gold:19]: This should be fixed in a future PR.  Should include the implicit scheme.

File: tests/gold_tests/pluginTest/tsapi/log.gold [hunk 3]
```diff
@@ -1,14 +1,87 @@
 Global: event=TS_EVENT_HTTP_TXN_START
 Global: event=TS_EVENT_HTTP_READ_REQUEST_HDR
 TSHttpTxnEffectiveUrlStringGet():  http://mYhOsT.teSt:SERVER_PORT/
+Client Request:
 TSHttpHdrEffectiveUrlBufGet():  http://myhost.test:SERVER_PORT/
+TSUrlSchemeGet():  http
+TSUrlRawSchemeGet():  http
+TSUrlPortGet():  80
```
Comment (ywkaras) [inline tests/gold_tests/pluginTest/tsapi/log.gold:8]: This should return SERVER_PORT, to be fixed in future PR.

File: tests/gold_tests/pluginTest/tsapi/tsapi.test.py [hunk 4]
```diff
@@ -71,7 +71,15 @@
 tr.Processes.Default.StartBefore(Test.Processes.ts)
 #
 tr.Processes.Default.Command = (
-    'curl --verbose --ipv4 --header "Host: mYhOsT.teSt:{0}" hTtP://loCalhOst:{1}/'.format(server.Variables.Port, ts.Variables.port)
+    r'printf "GET / HTTP/1.1\r\nHost: mYhOsT.teSt:{0}\r\n\r\n" | nc localhost {1}'.format(
```
Comment (randall) [inline tests/gold_tests/pluginTest/tsapi/tsapi.test.py:74]: Why nc?

Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py:74]: To me it's more clear what the exact contents of the HTTP message is when I use nc rather than curl.

**Review by SolidWallOfCode — APPROVED** [review]

File: tests/gold_tests/pluginTest/tsapi/tsapi.test.py (hunk 4 above)
Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: I prefer consistently using curl for all HTTP versions as long as curl can do the same thing, because I can clearly see the tests do the same things on different versions.
If you think we all should use printf and nc for H1 test cases, you should make another PR to make the change for all H1 tests, and let's see what other people think. I don't think this change has to be done on this PR.

**Review by maskit — CHANGES_REQUESTED** [review]
I don't think using printf and nc instead of curl is a good idea. Can you keep the command line as is on this PR and make another PR for it?

File: tests/gold_tests/pluginTest/tsapi/tsapi.test.py (hunk 4 above)
Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: @bryancall what is your thinking about this? @d2r ?

I have one test run with the scheme and host in the URL, and one with just the path in the URL (with the host in a host header and no explicit scheme).  Have you seen how long the manpage for curl is?  Does it even say whether or not the scheme/host are in the request URL?  If I have to revisit this test case at some point in the future, I'd rather not have to have to dig into all the quirks of curl to understand it, given that there is a more straight-forward alternative.

Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: I'd say "Have you seen how long the H2 spec is?". I would definitely not make a binary data and send it with "openssl s_client" for this test.

Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: How do I control whether or not curl puts the scheme and host in the HTTP Request URL?

Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: `curl -x http://yourats/ http://foo/` puts them in RequestURL.

Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: What do you mean? -x is the abbreviation of --proxy.  What does that have to do with controlling whether or not the scheme and host are in the URL in the HTTP request?

Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: https://tools.ietf.org/html/rfc7230#section-5.3.2
> When making a request to a proxy, other than a CONNECT or server-wide
   OPTIONS request (as detailed below), a client MUST send the target
   URI in absolute-form as the request-target.

Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: I believe that, under the terminology of the RFC, the term "proxy" only applies to forward proxies.  A reverse proxy is called a "gateway".  I don't think the restriction you reference applies to gateways.

Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: If you use --proxy, curl thinks the server is a forward proxy regardless of your server configuration, and it makes the RequestURL you need.

Comment (ywkaras) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: OK I think I see what you mean now:
```
$ curl --verbose http://httpbin.org/ip 2>&1 | grep '^> '
> GET /ip HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.66.0-DEV
> Accept: */*
> 
$ curl --verbose -x httpbin.org:80 http://httpbin.org/ip 2>&1 | grep '^> '
> GET http://httpbin.org/ip HTTP/1.1
> Host: httpbin.org
> User-Agent: curl/7.66.0-DEV
> Accept: */*
> Proxy-Connection: Keep-Alive
> 
$
```
But to me this makes the test harder rather than easier to understand and maintain.  Notice that curl inserts several headers that are not relevant to the test.

Comment (maskit) [reply tests/gold_tests/pluginTest/tsapi/tsapi.test.py]: > Notice that curl inserts several headers that are not relevant to the test.

I think they are relevant because they may change ATS's behavior. I guess major clients send Proxy-Client when they use absolute form, so testing the behavior without the header may be missing the point. And we can easily make natural requests by using curl. I don't think I can always do it manually.

**Comment by ywkaras** [comment]
Hopefully we can discuss tools for generating test requests during the summit.  We had a long discussion about this in our (Vz) weekly meeting.

**Comment by ywkaras** [comment]
@SolidWallOfCode @shinrich what is your opinions as to whether we should prohibit the use of netcat in new Au tests?  I think curl is not generally the best option for testing, as it doesn't allow one to exactly and explicitly (thus readably) specify the contents of the HTTP request.  netcat and openssl s_client allow you to do this (in conjunction with bash printf/echo -e).  The only case where there is no alternative to curl seems to be for H2.

**Comment by ywkaras** [comment]
OK @maskit  no more netcat.

**Comment by maskit** [comment]
Now I don't see a test for absolute form request URL.

**Review by maskit — APPROVED** [review]
